pkg/middleware/stored_session.go

```diff
@@ -0,0 +1,165 @@
+		refreshPeriod:                      opts.RefreshPeriod,
+		refreshSessionWithProviderIfNeeded: opts.RefreshSessionIfNeeded,
+		validateSessionState:               opts.ValidateSessionState,
+	}
+	return ss.loadSession
+}
+
+// storedSessionLoader is responsible for loading sessions from cookie
+// identified sessions in the session store.
+type storedSessionLoader struct {
+	store                              sessionsapi.SessionStore
+	refreshPeriod                      time.Duration
+	refreshSessionWithProviderIfNeeded func(context.Context, *sessionsapi.SessionState) (bool, error)
+	validateSessionState               func(context.Context, *sessionsapi.SessionState) bool
+}
+
+// loadSession attempts to load a session as identified by the request cookies.
+// If no session is found, the request will be passed to the nex handler.
+// If a session was loader by a previous handler, it will not be replaced.
+func (s *storedSessionLoader) loadSession(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(rw http.ResponseWriter, req *http.Request) {
+		scope := GetRequestScope(req)
+		// If scope is nil, this will panic.
+		// A scope should always be injected before this handler is called.
+		if scope.Session != nil {
+			// The session was already loaded, pass to the next handler
+			next.ServeHTTP(rw, req)
+			return
+		}
+
+		session, err := s.getValidatedSession(rw, req)
+		if err != nil {
+			// In the case when there was an error loading the session,
+			// we should clear the session
+			logger.Printf("Error loading cookied session: %v, removing session", err)
+			s.store.Clear(rw, req)
+		}
+
+		// Add the session to the scope if it was found
+		scope.Session = session
+		next.ServeHTTP(rw, req)
+	})
+}
+
+// getValidatedSession is responsible for loading a session and making sure
+// that is is valid.
+func (s *storedSessionLoader) getValidatedSession(rw http.ResponseWriter, req *http.Request) (*sessionsapi.SessionState, error) {
+	session, err := s.store.Load(req)
+	if err != nil {
+		return nil, err
+	}
+	if session == nil {
+		// No session was found in the storage, nothing more to do
+		return nil, nil
+	}
+
+	err = s.refreshSessionIfNeeded(rw, req, session)
+	if err != nil {
+		return nil, fmt.Errorf("error refreshing access token for session (%s): %v", session, err)
+	}
+
+	return session, nil
+}
+
+// refreshSessionIfNeeded will attempt to refresh a session if the session
+// is older than the refresh period.
+// It is assumed that if the provider refreshes the session, the session is now
+// valid.
+// If the session requires refreshing but the provider does not refresh it,
+// we must validate the session to ensure that the returned session is still
+// valid.
+func (s *storedSessionLoader) refreshSessionIfNeeded(rw http.ResponseWriter, req *http.Request, session *sessionsapi.SessionState) error {
+	if s.refreshPeriod <= time.Duration(0) || session.Age() < s.refreshPeriod {
```

Comment:
So upon reflecting on this functionality... I think this is potentially not doing what we want (both legacy and refactor, so we can handle this in a different PR):

If I set `--cookie-refresh` it will only actually use my Refresh Token to get me a new Access/ID Token if those are expired.

So a `cookie-refresh` with a timer < IDToken expiry is essentially pointless (besides just enabling refreshing in general) -- But I think that gets us into trouble on the edges of an IDToken expiration time. If we have clock skew of a request comes in right at expiration and validates in oauth2-proxy -- by the time a backend reads it via `pass-authorization` it might be expired.

I would think users would want to set `cookie-refresh` < token TTL to avoid that race and emulate logic like this that Kubernetes does with projected service account tokens (they auto-refresh the mount with a new JWT at 80% of lifespan): https://kubernetes.io/docs/tasks/configure-pod-container/configure-service-account/#service-account-token-volume-projection

That way you always have a valid ID/Access token and never have to worry about clock drift passing the token to the backend.

I guess the only current viable use of `--cookie-refresh` is to set a refresh time > IDToken lifespan to delay/skip the IDToken refresh...? Or refresh time == IDToken lifespan just to enable refreshing in general?

Reply:
Normally I've recommended setting it to the ID Token expiry time. But you're right, I'm also aware that this logic needs looking at. Would be good to go back over the original PR that implemented refreshing to see what the logic was then, we may have drifted away from it at some point.

I suspect the intention was that on refresh, we should update the `createdAt` every time so that the session is actually refreshed somehow at least. I'm aware that some people have reported bugs around this and saying they just get loops of refreshing old session, so we have some work to do around it, as you say, separate PR for this I reckon.